doc/manual/contract.md

```diff
@@ -0,0 +1,505 @@
+
+You can put contracts on different elements of the language: methods, attributes, interfaces and classes in order to verify the conformity of the implementation compared with the specification.
+Different types of contracts exist. The preconditions (`expect`) and the postconditions (`ensure`) can be put on methods and attributes (setter) to define respectively the condition to check before and after the execution of the method. The invariants (`invariant`) that can be put on class and interface to define the conditions that must remain unchanged before and after the execution of each method of the class.
+
+The implementation of contracts is recommended when we are dealing with business rules. For example when we are developing a library it may be necessary to specify contracts to guarantee valid usages of wer API. Also, contracts support inheritance, which guarantees that contracts will be satisfied in future specification.
+
+Let us take a look at the `Stack` class bellow:
+
+~~~
+class Stack[T]
+	invariant(size >= 0)
+
+	var data = new Array[T]
+
+	var size = 0
+
+	fun is_empty: Bool
+	is
+		ensure(result == (size == 0))
+	do
+		return size == 0
+	end
+
+	# Add `item` on top of the stack
+	fun push(item: T)
+	is
+		ensure(size == old(size) + 1)
+		ensure(data[size-1] == item)
+	do
+		data[size] = item
+		size += 1
+	end
+
+	# Remove the item on top of the stack
+	fun pop
+	is
+		expect(not is_empty)
+		ensure(size == old(size) - 1)
+	do
+		data.pop
+		size -= 1
+	end
+
+	# Return the item at the given index.
+	fun [](index: Int): T
+	is
+		expect(index < size)
+		expect(index >= 0)
+		ensure(result == data[index])
+	do
+		return data[index]
+	end
+
+	# Set the item at the given index
+	fun []=(index: Int, item: T)
+	is
+		expect(index >= 0)
+		expect(index <= size)
+	do
+		data[index] = item
+	end
+end
+~~~
+
+As we can see the contracts are defined using several annotations: `invariant` which designates the class invariants, `expect` which designates the preconditions and `ensure` which designates the postconditions. The conditions of a contract must be expressed using boolean expressions in the same way as assertions. Each contract of this example will be detailed in the sections below.
+
+By default when you compile or interpret (`nitc stack.nit` or `nit stack.nit`) a program, all contracts (`ensure`, `expect`, `invariant`) used in the main module are enabled, direct imported modules have only the expects contracts enabled (`ensure` and `invariant` contracts are disabled). The other indirect imported modules don't have active contracts. We can use options to change the contract rule see `Options` section for more information.
```

Comment:
I think it will be easier to understand if we can related to the current example.

What If I compile:
* `stack.nit`
* `my_file_that_uses_stack.nit`

Also what about the stdlib? is it directly or indirectly imported here?

Reply:
Paragraphs that list things can be broken into bullet points to make them easier to scan/read.